### ISSUE TITLE
Mask edge or node during training only to do on the fly graph augmentation

### ISSUE THREAD

**Issue by thegodone** [akensert/molgraph#6]
I would like to implement this code in your toolkit: https://github.com/BaratiLab/AugLiChem/tree/main/auglichem/molecule

I see that could be a useful addition to your module ?

https://github.com/BaratiLab/AugLiChem/blob/4ac7a682de49d8f2b9ce6d6551c18b479f6208bd/docs/molecule.md?plain=1#L68 (Atom and Bond mask looks interesting based on paper results).

**Comment by thegodone** [comment]
I am at this stage so far: 

```
class AtomMaskingLayer(tf.keras.layers.Layer):
    def __init__(self, masking_ratio, **kwargs):
        super(AtomMaskingLayer, self).__init__(**kwargs)
        self.masking_ratio = masking_ratio

    def call(self, inputs, training=None):
        if training:
            graph_tensor = inputs  
            node_feature = graph_tensor.node_feature

            num_atoms = tf.shape(node_feature)[0]
            num_masked_atoms = tf.cast(self.masking_ratio * tf.cast(num_atoms, tf.float32), tf.int32)

            mask_indices = tf.random.uniform((num_masked_atoms,), maxval=num_atoms, dtype=tf.int32)
            mask = tf.reduce_sum(tf.one_hot(mask_indices, depth=num_atoms), axis=0)

            masked_node_feature = node_feature * (1 - tf.reshape(mask, [-1, 1]))
            graph_tensor = graph_tensor.update({'node_feature': masked_node_feature})

            return graph_tensor
        else:
            return inputs

```
Do you validate the behaviour ?

**Comment by thegodone** [comment]
Goal is to have training augmentation by minimal random variation of the graph object. This was validated in recent papers. I confirm that it help to avoid overfit of training dataset and so found stronger knowledge of molecules embedding.I read several a papers on graph autoencoder. Will send you a new issue soon sorry ;-)Envoyé de mon iPhoneLe 6 juin 2023 à 13:05, Alexander Kensert ***@***.***> a écrit :﻿
What is the purpose of the masking?
I'm actually working on implementing some models for pretraining, based on this paper.
I think your approach is valid, though I was planning on implementing masks on token level. So using molgraph.chemistry.Tokenizer instead of molgraph.chemistry.Featurizer, and then have a masking layer which adds mask_token to random nodes/edges. But I also like the idea of masking these vectors instead of tokens.

**Comment by akensert** [comment]
Okay, so you would mask random node/edge features at the beginning? What should the masked node/edge features be? In your case zero-vectors, right? Though zeros could have a meaning, I guess this is the best option?

**Comment by thegodone** [comment]
I need to mask in the fly so at each epoch molecule is very slightly different enougth to avoid memorizing. And yes ideally we need to remove the connectivity instead of masking. Because I am not fully confident with your graphTensor, that was my first attempt. In the github source code upper, they indeed remove the atom or bond randomly in a dynamic way and can combine the manipulations.Envoyé de mon iPhoneLe 6 juin 2023 à 13:15, Alexander Kensert ***@***.***> a écrit :﻿
Okay, so you would mask random node/edge features at the beginning? What should the masked node/edge features be? In your case zero-vectors, right? Though zeros could have a meaning, I guess this is the best option?

**Comment by thegodone** [comment]
Correct max freq is fine.Envoyé de mon iPhoneLe 6 juin 2023 à 13:24, Alexander Kensert ***@***.***> a écrit :﻿
I will try to implement a layer like that.
So completely dropping nodes and edges at random, with a frequency of mask_freq ?

**Comment by thegodone** [comment]
looks good yes

**Comment by akensert** [comment]
As you can see it was pretty non-trivial to perform the node masking. If the graph is huge, it could consume a lot of memory:

``python
tf.expand_dims(tensor.edge_dst, -1) == keep_nodes
```

An option is to make it ragged and "loop" over each subgraph. Via `tf.map_fn`. 

Perhaps something to add later. If the graph is of a certain size, it "loops", otherwise everything at once.

**Comment by akensert** [comment]
@thegodone I'm going to change the name of the class. `FeatureMasking` is not appropriate as complete nodes (and/or edges) are removed; and not simply the features.

**Comment by akensert** [comment]
Instead of `FeatureMasking` there will be a `NodeMasking` and `EdgeMasking`, each of which completely removes nodes and edges from the graph respectively.

**Comment by akensert** [comment]
@thegodone could you possibly link me the paper(s)? :) 

Regarding: _"Goal is to have training augmentation by minimal random variation of the graph object. This was validated in recent papers. I confirm that it help to avoid overfit of training dataset and so found stronger knowledge of molecules embedding."_

**Comment by thegodone** [comment]
the link : https://dx.doi.org/10.1088/2632-2153/ac9c84
